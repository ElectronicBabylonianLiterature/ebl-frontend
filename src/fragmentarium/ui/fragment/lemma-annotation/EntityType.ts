const EntityTypes = {
  LOCATION: { type: 'LOCATION', label: 'LOC' },
  PERSON: { type: 'PERSON', label: 'PERSON' },
} as const
export type EntityType = keyof typeof EntityTypes
