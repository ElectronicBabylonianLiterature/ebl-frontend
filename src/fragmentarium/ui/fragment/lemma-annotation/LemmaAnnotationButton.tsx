import EditableToken from 'fragmentarium/ui/fragment/linguistic-annotation/EditableToken'
import React from 'react'
import { Button, ButtonGroup, Dropdown } from 'react-bootstrap'
import DisplayToken from 'transliteration/ui/DisplayToken'

export interface LemmaActionCallbacks {
  onResetCurrent: () => void
  onMouseEnter: () => void
  onMouseLeave: () => void
  onMultiApply: () => void
  onMultiReset: () => void
}

export default function LemmaActionButton({
  token,
  onResetCurrent,
  onMouseEnter,
  onMouseLeave,
  onMultiApply,
  onMultiReset,
}: {
  token: EditableToken
} & LemmaActionCallbacks): JSX.Element {
  return (
    <Dropdown as={ButtonGroup} className="lemmatizer__editor__action-button">
      <Button
        variant="secondary"
        onClick={onResetCurrent}
        disabled={!token.isDirty}
        aria-label="reset-current-token"
      >
        <i className={'fas fa-rotate-left'}></i>
      </Button>

      <Dropdown.Toggle split variant="secondary" id="dropdown-split-basic">
        <i className={'fas fa-caret-down'}></i>
      </Dropdown.Toggle>

      <Dropdown.Menu>
        <Dropdown.Item
          onMouseEnter={onMouseEnter}
          onMouseLeave={onMouseLeave}
          onClick={onMultiApply}
        >
          Update all instances of <DisplayToken token={token.token} />
        </Dropdown.Item>
        <Dropdown.Divider />
        <Dropdown.Item
          onMouseEnter={onMouseEnter}
          onMouseLeave={onMouseLeave}
          onClick={onMultiReset}
          disabled={!token.isDirty}
        >
          Reset all instances of <DisplayToken token={token.token} />
        </Dropdown.Item>
      </Dropdown.Menu>
    </Dropdown>
  )
}
