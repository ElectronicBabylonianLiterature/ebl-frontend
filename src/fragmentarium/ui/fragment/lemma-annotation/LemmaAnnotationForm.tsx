import WordService from 'dictionary/application/WordService'
import EditableToken from 'fragmentarium/ui/fragment/linguistic-annotation/EditableToken'
import React from 'react'
import AsyncSelect from 'react-select/async'
import { ActionMeta, ValueType } from 'react-select'
import {
  LemmaOption,
  Option,
  MultiValueLabel,
} from 'fragmentarium/ui/lemmatization/LemmaSelectionForm'

type Props = {
  token: EditableToken | null
  wordService: WordService
  onChange: (options: LemmaOption[] | null) => void
  onTab: () => void
  onShiftTab: () => void
}
type State = {
  options: ValueType<LemmaOption, true>
  menuIsOpen?: boolean
}

export default class LemmaAnnotationForm extends React.Component<Props, State> {
  private token: EditableToken | null

  constructor(props: Props) {
    super(props)

    this.token = props.token

    this.state = {
      options: this.token?.lemmas,
      menuIsOpen: undefined,
    }
  }

  loadOptions = (
    inputValue: string,
    callback: (lemmas: LemmaOption[]) => void
  ): void => {
    this.props.wordService
      .searchLemma(inputValue)
      .then((words) => words.map((word) => new LemmaOption(word)))
      .then(callback)
  }

  handleChange = (
    options: ValueType<LemmaOption, true>,
    { action, removedValue }: ActionMeta<LemmaOption>
  ): void => {
    const current = this.state.options || []

    const update =
      action === 'remove-value'
        ? current.filter((option) => option.id !== removedValue?.id)
        : options

    this.setState({
      ...this.state,
      options: update,
    })

    this.props.onChange((update as LemmaOption[]) || null)
  }

  handleKeyDown = (event: React.KeyboardEvent): void => {
    if (event.code === 'Tab') {
      event.preventDefault()
      if (event.shiftKey) {
        this.props.onTab()
      } else {
        this.props.onShiftTab()
      }
    }
  }

  render(): JSX.Element {
    return (
      <AsyncSelect
        aria-label="edit-token-lemmas"
        className="lemmatizer__editor__lemma-select"
        autoFocus={true}
        isDisabled={!this.token?.isSelected}
        isClearable={false}
        isMulti={true}
        isSearchable={true}
        loadOptions={this.loadOptions}
        onChange={this.handleChange}
        onKeyDown={this.handleKeyDown}
        menuIsOpen={this.state.menuIsOpen}
        value={this.state.options}
        placeholder={'---'}
        components={{ Option, MultiValueLabel }}
      />
    )
  }
}
