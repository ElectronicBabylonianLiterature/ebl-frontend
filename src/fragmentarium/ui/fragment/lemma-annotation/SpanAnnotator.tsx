import {
  entities,
  EntityAnnotationSpan,
  EntityType,
} from 'fragmentarium/ui/fragment/lemma-annotation/EntityType'
import AnnotationContext from 'fragmentarium/ui/fragment/lemma-annotation/TextAnnotationContext'
import _ from 'lodash'
import React, { useContext } from 'react'
import Select from 'react-select'

export function clearSelection(): void {
  if (window.getSelection) {
    if (window.getSelection()?.empty) {
      window.getSelection()?.empty()
    } else if (window.getSelection()?.removeAllRanges) {
      window.getSelection()?.removeAllRanges()
    }
  }
}

interface EntityTypeOption {
  label: EntityType
  value: EntityType
}

const options: EntityTypeOption[] = entities.map((entity) => ({
  value: entity.type,
  label: entity.type,
}))

function createId(
  type: EntityType,
  annotationSpans: readonly EntityAnnotationSpan[]
): string {
  const currentMaxId =
    _.max(
      annotationSpans
        .filter((entity) => entity.type === type)
        .map(({ id }) => parseInt(id.split('-')[1]))
    ) || 0

  return `E-${currentMaxId + 1}`
}

export default function SpanAnnotator({
  selection,
  setSelection,
}: {
  selection: readonly string[]
  setSelection: React.Dispatch<React.SetStateAction<readonly string[]>>
}): JSX.Element {
  const [
    selectedType,
    setSelectedType,
  ] = React.useState<EntityTypeOption | null>(null)
  const [entities, dispatch] = useContext(AnnotationContext)

  return (
    <div onMouseUp={(event) => event.stopPropagation()}>
      <Select
        autoFocus
        options={options}
        value={selectedType}
        onChange={(option) => {
          setSelectedType(option as EntityTypeOption)
          if (option) {
            const entity: EntityAnnotationSpan = {
              id: createId(option.value, entities),
              type: option.value,
              span: selection,
              tier: 1,
            }
            dispatch({ type: 'add', entity })
            clearSelection()
            setSelection([])
          }
        }}
      />
    </div>
  )
}
