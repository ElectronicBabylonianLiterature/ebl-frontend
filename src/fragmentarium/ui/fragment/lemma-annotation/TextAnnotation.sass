$highlight-color: lightsteelblue
$space-width: .25rem

@mixin subscript-label
  font-size: 75%
  font-weight: bold
  font-family: monospace
  line-height: 0
  position: relative
  vertical-align: baseline
  bottom: -0.4em
  padding-left: .1rem
  padding-right: .1rem

@mixin highlight-chain($depth: 1, $color: $highlight-color)
  & + span:not(.markable)
    background: $color

    @if $depth > 1
      @include highlight-chain($depth - 1, $color)

@mixin ensure-black-font
  span[class^="Transliteration__"]
    color: black !important

.entity-span
  @include ensure-black-font

.text-annotation
  &__line-number
    ::selection
      background-color: transparent !important
      color: inherit

.markable
  position: relative
  &.selected
    background: $highlight-color

    &:not(.span-end)
      @include highlight-chain(3)

    @include ensure-black-font

.entity-marker
  height: 4px
  background-color: sandybrown

  &:not(.span-end)
    width: calc(100% + .5em)

@mixin indicator-label($label, $color)
  background-color: $color

  &.initial::before
    content: "#{$label}"
    font-size: 50%
    font-weight: bold
    font-family: monospace
    padding: .1rem .25rem
    background-color: $color
    border-radius: .25rem
    position: absolute
    left: 0
    top: 50%
    transform: translateY(-50%)
    z-index: 1

.span-indicator
  width: calc(100% + #{$space-width})
  position: absolute
  left: 0
  height: 3px

  &.final
    width: 100%

.named-entity
  &__LOCATION
    @include indicator-label(LOC, sandybrown)

  &__PERSON
    @include indicator-label(PERSON,mediumseagreen)

tr:has(.span-indicator) > td
  padding-bottom: 2.5rem
