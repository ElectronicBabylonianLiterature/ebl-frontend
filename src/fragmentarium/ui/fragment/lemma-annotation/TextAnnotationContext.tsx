import React, { Dispatch, useReducer } from 'react'
import { EntityType } from 'fragmentarium/ui/fragment/lemma-annotation/SpanAnnotator'

export interface Entity {
  id: string
  span: readonly string[]
  type: EntityType
}

type State = readonly Entity[]
export type AnnotationContextService = [State, Dispatch<Action>]

type AddAction = {
  type: 'add'
  entity: Entity
}

export type Action = AddAction

const AnnotationContext = React.createContext<AnnotationContextService>([
  [],
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  () => {},
])

function reducer(state: State, action: Action): State {
  switch (action.type) {
    case 'add':
      return [...state, action.entity]
  }
}

export function useAnnotationContext(
  initial: Entity[] = []
): AnnotationContextService {
  return useReducer(reducer, initial)
}

export default AnnotationContext
