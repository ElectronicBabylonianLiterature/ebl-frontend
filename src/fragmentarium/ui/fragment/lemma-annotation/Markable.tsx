import React, { PropsWithChildren, useContext } from 'react'
import { AnyWord } from 'transliteration/domain/token'
import './TextAnnotation.sass'
import classNames from 'classnames'
import _ from 'lodash'
import { OverlayTrigger, Popover } from 'react-bootstrap'
import SpanAnnotator, {
  clearSelection,
} from 'fragmentarium/ui/fragment/lemma-annotation/SpanAnnotator'
import { EntityAnnotationSpan } from 'fragmentarium/ui/fragment/lemma-annotation/EntityType'
import AnnotationContext from 'fragmentarium/ui/fragment/lemma-annotation/TextAnnotationContext'

const markableClass = 'markable'

function sortSelection(
  selection: readonly string[],
  words: readonly string[]
): readonly string[] {
  return _.sortBy(selection, (id) => words.indexOf(id))
}

function expandSelection(
  start: string,
  end: string,
  words: readonly string[]
): readonly string[] {
  const positions = [words.indexOf(start), words.indexOf(end)]
  const [startIndex, endIndex] = _.sortBy(positions)

  return words.slice(startIndex, endIndex + 1)
}

function getTokenId(node: Node | null): string | null {
  const tokenNode = node?.parentElement?.closest(`.${markableClass}`)
  return tokenNode ? tokenNode.getAttribute('data-id') : null
}

function getSelectedTokens(words: readonly string[]): readonly string[] {
  const selection = document.getSelection()
  if (selection) {
    const start = getTokenId(selection.anchorNode)
    const end = getTokenId(selection.focusNode)

    if (start && end) {
      clearSelection()
      return expandSelection(start, end, words)
    }
  }
  return []
}

function isSelected(token: AnyWord, selection: readonly string[]): boolean {
  return !!token.id && selection.includes(token.id)
}

function mergeSelections(
  selection: readonly string[],
  newSelection: readonly string[]
): readonly string[] {
  return _.isEmpty(_.intersection(selection, newSelection))
    ? _.union(selection, newSelection)
    : _.difference(selection, newSelection)
}

function SpanIndicator({
  tokenId,
  entitySpan,
  hoveredSpanId,
  setHoveredSpanId,
}: {
  tokenId?: string
  entitySpan: EntityAnnotationSpan
  hoveredSpanId: string | null
  setHoveredSpanId: React.Dispatch<React.SetStateAction<string | null>>
}): JSX.Element {
  const isInitial = tokenId === _.first(entitySpan.span)
  return (
    <span
      onMouseEnter={() => setHoveredSpanId(entitySpan.id)}
      onMouseLeave={() => setHoveredSpanId(null)}
      onMouseUp={() => {
        console.log(`clicked on ${entitySpan.id}`)
      }}
      data-span-id={entitySpan.id}
      className={classNames(
        'span-indicator',
        `tier-depth--${entitySpan.tier}`,
        `named-entity__${entitySpan.type}`,
        {
          hovered: hoveredSpanId === entitySpan.id,
          initial: isInitial,
          final: tokenId === _.last(entitySpan.span),
        }
      )}
    />
  )
}

export default function Markable({
  token,
  words,
  selection,
  setSelection,
  hoveredSpanId,
  setHoveredSpanId,
  children,
}: PropsWithChildren<{
  token: AnyWord
  words: readonly string[]
  selection: readonly string[]
  setSelection: React.Dispatch<React.SetStateAction<readonly string[]>>
  hoveredSpanId: string | null
  setHoveredSpanId: React.Dispatch<React.SetStateAction<string | null>>
}>): JSX.Element {
  const [{ entities }] = useContext(AnnotationContext)

  function handleSelection(event: React.MouseEvent) {
    const newSelection = getSelectedTokens(words)

    setSelection(
      sortSelection(
        event.altKey ? mergeSelections(selection, newSelection) : newSelection,
        words
      )
    )

    event.stopPropagation()
  }

  const popover = (
    <Popover id={_.uniqueId('SpanAnnotationPopOver-')}>
      <Popover.Title>{`Annotate ${selection.length} tokens`}</Popover.Title>
      <Popover.Content>
        <SpanAnnotator selection={selection} setSelection={setSelection} />
      </Popover.Content>
    </Popover>
  )

  return (
    <span
      className={classNames(markableClass, {
        selected: isSelected(token, selection),
        'span-end': token.id === _.last(selection),
      })}
      data-id={token.id}
    >
      <OverlayTrigger
        trigger={['click']}
        overlay={popover}
        placement={'top'}
        show={!!token.id && _.head(selection) === token.id}
      >
        <span onMouseUp={handleSelection}>{children}</span>
      </OverlayTrigger>
      {entities.map((entity, index) => {
        return token.id && entity.span.includes(token.id) ? (
          <SpanIndicator
            key={index}
            tokenId={token.id}
            entitySpan={entity}
            hoveredSpanId={hoveredSpanId}
            setHoveredSpanId={setHoveredSpanId}
          />
        ) : (
          <React.Fragment key={index} />
        )
      })}
    </span>
  )
}
