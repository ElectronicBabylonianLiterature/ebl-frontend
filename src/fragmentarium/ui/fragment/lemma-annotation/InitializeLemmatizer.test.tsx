import React from 'react'
import { render, screen } from '@testing-library/react'
import Promise from 'bluebird'
import { InitializeLemmatizer } from 'fragmentarium/ui/fragment/lemma-annotation/InitializeLemmatizer'
import { Text } from 'transliteration/domain/text'
import { AbstractLine } from 'transliteration/domain/abstract-line'
import WordService from 'dictionary/application/WordService'
import FragmentService from 'fragmentarium/application/FragmentService'
import Word from 'dictionary/domain/Word'
import { act } from 'react-dom/test-utils'
import textLine from 'test-support/lines/text-line'
import { TextLine } from 'transliteration/domain/text-line'
import { AkkadianWord } from 'transliteration/domain/token'
import { wordFactory } from 'test-support/word-fixtures'

const token: AkkadianWord = {
  enclosureType: [],
  erasure: 'NONE',
  cleanValue: 'kur',
  value: 'kur',
  language: 'AKKADIAN',
  normalized: true,
  lemmatizable: true,
  alignable: true,
  uniqueLemma: ['aklu I'],
  alignment: null,
  variant: null,
  modifiers: [],
  parts: [
    {
      enclosureType: [],
      erasure: 'NONE',
      cleanValue: 'kur',
      value: 'kur',
      name: 'kur',
      nameParts: [
        {
          enclosureType: [],
          erasure: 'NONE',
          cleanValue: 'kur',
          value: 'kur',
          type: 'ValueToken',
        },
      ],
      subIndex: 1,
      modifiers: [],
      flags: [],
      sign: null,
      type: 'Reading',
    },
  ],
  type: 'AkkadianWord',
  hasVariantAlignment: false,
  hasOmittedAlignment: false,
}
const partlyLemmatizedLine = new TextLine({
  prefix: '2.',
  content: [token],
  lineNumber: {
    number: 2,
    hasPrime: false,
    prefixModifier: null,
    suffixModifier: null,
    type: 'LineNumber',
  },
  type: 'TextLine',
})

const lines: readonly AbstractLine[] = [textLine, partlyLemmatizedLine]
const text = new Text({ lines })
const words: Word[] = [
  wordFactory.build({
    _id: 'aklu I',
    lemma: ['aklu'],
    homonym: 'I',
  }),
]
const fragmentNumber = 'Test.Fragment'
let container

jest.mock('dictionary/application/WordService')
jest.mock('fragmentarium/application/FragmentService')

const MockFragmentService = FragmentService as jest.Mock<
  jest.Mocked<FragmentService>
>
const fragmentServiceMock = new MockFragmentService()
const MockWordService = WordService as jest.Mock<jest.Mocked<WordService>>
const wordServiceMock = new MockWordService()
const updateLemmaAnnotation = jest.fn()

beforeEach(async () => {
  wordServiceMock.findAll.mockReturnValue(Promise.resolve(words))
  await act(async () => {
    container = await render(
      <InitializeLemmatizer
        text={text}
        museumNumber={fragmentNumber}
        wordService={wordServiceMock}
        fragmentService={fragmentServiceMock}
        updateAnnotation={updateLemmaAnnotation}
      />
    ).container
  })
})

it('displays the annotation tool', async () => {
  expect(container).toMatchSnapshot()
})
it('displays the existing lemmas', async () => {
  expect(screen.getByText('aklu I')).toBeVisible()
})
it('loads existing lemmas', async () => {
  expect(wordServiceMock.findAll).toBeCalledWith(['aklu I'])
})
