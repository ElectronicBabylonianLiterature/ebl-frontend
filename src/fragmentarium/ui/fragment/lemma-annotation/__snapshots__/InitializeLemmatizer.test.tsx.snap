// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays the annotation tool 1`] = `
<div>
  <div
    class="lemmatizer__anno-tool container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="modal show lemmatizer__editor"
        >
          <div
            class="modal-dialog lemmatizer__modal"
          >
            <div
              class="modal-content"
            >
              <div
                class="modal-header"
              >
                <h6
                  class="modal-title"
                >
                  1: kur
                </h6>
              </div>
              <div
                class="modal-body"
              >
                <form
                  class=""
                >
                  <div
                    class="lemmatizer__editor__row form-group row"
                  >
                    <div
                      class="lemmatizer__editor__lemma-select css-2b097c-container"
                    >
                      <span
                        aria-live="polite"
                        class="css-1laao21-a11yText"
                      >
                        <span
                          id="aria-selection-event"
                        >
                           
                          
                        </span>
                        <span
                          id="aria-context"
                        >
                           
                            0 results available. edit-token-lemmas is focused ,type to refine list, press Down to open the menu,  press left to focus selected values
                        </span>
                      </span>
                      <div
                        class=" css-1pahdxg-control"
                      >
                        <div
                          class=" css-g1d714-ValueContainer"
                        >
                          <div
                            class=" css-1wa3eu0-placeholder"
                          >
                            ---
                          </div>
                          <div
                            class="css-b8ldur-Input"
                          >
                            <div
                              class=""
                              style="display: inline-block;"
                            >
                              <input
                                aria-autocomplete="list"
                                aria-label="edit-token-lemmas"
                                autocapitalize="none"
                                autocomplete="off"
                                autocorrect="off"
                                id="react-select-2-input"
                                spellcheck="false"
                                style="box-sizing: content-box; width: 2px; border: 0px; opacity: 1; outline: 0; padding: 0px;"
                                tabindex="0"
                                type="text"
                                value=""
                              />
                              <div
                                style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                              />
                            </div>
                          </div>
                        </div>
                        <div
                          class=" css-1hb7zxy-IndicatorsContainer"
                        >
                          <span
                            class=" css-1okebmr-indicatorSeparator"
                          />
                          <div
                            aria-hidden="true"
                            class=" css-1gtu0rj-indicatorContainer"
                          >
                            <svg
                              aria-hidden="true"
                              class="css-6q0nyr-Svg"
                              focusable="false"
                              height="20"
                              viewBox="0 0 20 20"
                              width="20"
                            >
                              <path
                                d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                              />
                            </svg>
                          </div>
                        </div>
                      </div>
                    </div>
                    <div
                      class="lemmatizer__editor__action-button dropdown btn-group"
                      role="group"
                    >
                      <button
                        aria-label="reset-current-token"
                        class="btn btn-secondary"
                        disabled=""
                        type="button"
                      >
                        <i
                          class="fas fa-rotate-left"
                        />
                      </button>
                      <button
                        aria-expanded="false"
                        aria-haspopup="true"
                        class="dropdown-toggle dropdown-toggle-split btn btn-secondary"
                        disabled=""
                        id="dropdown-split-basic"
                        type="button"
                      >
                        <i
                          class="fas fa-caret-down"
                        />
                      </button>
                    </div>
                  </div>
                </form>
              </div>
              <div
                class="lemmatizer__editor__footer modal-footer"
              >
                <button
                  aria-label="autofill-lemmas"
                  class="btn btn-outline-primary"
                  type="button"
                >
                  <i
                    class="fas fa-wand-magic-sparkles"
                  />
                   
                  Autofill
                </button>
                <button
                  aria-label="save-updates"
                  class="btn btn-primary"
                  disabled=""
                  type="button"
                >
                  Save
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="lemmatizer__text-col col"
      >
        <div
          class="lemmatizer__text-wrapper"
        >
          <table
            class="Transliteration__lines"
          >
            <tbody>
              <tr
                id="line-1"
              >
                <td
                  class="Transliteration__TextLine"
                >
                  <sup>
                    (
                    1
                    )
                  </sup>
                </td>
                <td
                  colspan="1"
                >
                  <span
                    class="Transliteration__Word Transliteration__Word--AKKADIAN"
                  >
                    <span
                      class=""
                    >
                      <span
                        class="Transliteration__Reading"
                      >
                        <span
                          class=""
                        >
                          <span
                            class="Transliteration__ValueToken"
                          >
                            <span
                              class=""
                            >
                              érin
                            </span>
                          </span>
                          <sup
                            class="Transliteration__modifier"
                          />
                          <sup
                            class="Transliteration__flag"
                          />
                        </span>
                      </span>
                      <span
                        class="Transliteration__Joiner"
                      >
                        <span
                          class=""
                        >
                          -
                        </span>
                      </span>
                      <span
                        class="Transliteration__BrokenAway"
                      >
                        <span
                          class=""
                        >
                          [
                        </span>
                      </span>
                      <span
                        class="Transliteration__PerhapsBrokenAway"
                      >
                        <span
                          class="Transliteration__PerhapsBrokenAway--BROKEN_AWAY"
                        >
                          (
                        </span>
                      </span>
                      <span
                        class="Transliteration__Logogram"
                      >
                        <span
                          class="Transliteration__Logogram--BROKEN_AWAY Transliteration__Logogram--PERHAPS_BROKEN_AWAY"
                        >
                          <span
                            class="Transliteration__ValueToken"
                          >
                            <span
                              class="Transliteration__ValueToken--BROKEN_AWAY Transliteration__ValueToken--PERHAPS_BROKEN_AWAY"
                            >
                              ŠÁ
                            </span>
                          </span>
                          <sup
                            class="Transliteration__modifier"
                          />
                          <sup
                            class="Transliteration__flag"
                          />
                        </span>
                      </span>
                      <span
                        class="Transliteration__Joiner"
                      >
                        <span
                          class="Transliteration__Joiner--BROKEN_AWAY Transliteration__Joiner--PERHAPS_BROKEN_AWAY"
                        >
                          -
                        </span>
                      </span>
                      <span
                        class="Transliteration__Variant"
                      >
                        <span
                          class="Transliteration__Reading"
                        >
                          <span
                            class="Transliteration__Reading--BROKEN_AWAY Transliteration__Reading--PERHAPS_BROKEN_AWAY"
                          >
                            <span
                              class="Transliteration__ValueToken"
                            >
                              <span
                                class="Transliteration__ValueToken--BROKEN_AWAY Transliteration__ValueToken--PERHAPS_BROKEN_AWAY"
                              >
                                šà
                              </span>
                            </span>
                            <sup
                              class="Transliteration__modifier"
                            />
                            <sup
                              class="Transliteration__flag"
                            />
                          </span>
                        </span>
                        /
                        <span
                          class="Transliteration__Logogram"
                        >
                          <span
                            class="Transliteration__Logogram--BROKEN_AWAY Transliteration__Logogram--PERHAPS_BROKEN_AWAY"
                          >
                            <span
                              class="Transliteration__ValueToken"
                            >
                              <span
                                class="Transliteration__ValueToken--BROKEN_AWAY Transliteration__ValueToken--PERHAPS_BROKEN_AWAY"
                              >
                                GÌRI
                              </span>
                            </span>
                            <sup
                              class="Transliteration__modifier"
                            />
                            <sup
                              class="Transliteration__flag"
                            />
                          </span>
                        </span>
                      </span>
                      <span
                        class="Transliteration__PerhapsBrokenAway"
                      >
                        <span
                          class="Transliteration__PerhapsBrokenAway--BROKEN_AWAY Transliteration__PerhapsBrokenAway--PERHAPS_BROKEN_AWAY"
                        >
                          )
                        </span>
                      </span>
                    </span>
                  </span>
                  <span
                    class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
                  >
                     
                  </span>
                  <span
                    class="lemmatizer__token-wrapper editable selected"
                    role="button"
                  >
                    <span
                      class="Transliteration__Word Transliteration__Word--AKKADIAN"
                    >
                      <span
                        class=""
                      >
                        <span
                          class="Transliteration__Reading"
                        >
                          <span
                            class=""
                          >
                            <span
                              class="Transliteration__ValueToken"
                            >
                              <span
                                class=""
                              >
                                kur
                              </span>
                            </span>
                            <sup
                              class="Transliteration__modifier"
                            />
                            <sup
                              class="Transliteration__flag"
                            />
                          </span>
                        </span>
                      </span>
                    </span>
                    <span
                      class="badge badge-danger"
                    >
                      Empty
                    </span>
                  </span>
                  <span
                    class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
                  >
                     
                  </span>
                  <span
                    class="lemmatizer__token-wrapper editable"
                    role="button"
                  >
                    <span
                      class="Transliteration__Word Transliteration__Word--AKKADIAN"
                    >
                      <span
                        class="Transliteration__Word--BROKEN_AWAY"
                      >
                        <span
                          class="Transliteration__Reading"
                        >
                          <span
                            class="Transliteration__Reading--BROKEN_AWAY"
                          >
                            <span
                              class="Transliteration__ValueToken"
                            >
                              <span
                                class="Transliteration__ValueToken--BROKEN_AWAY"
                              >
                                saḫ
                              </span>
                            </span>
                            <sup
                              class="Transliteration__modifier"
                            />
                            <sup
                              class="Transliteration__flag"
                            />
                          </span>
                        </span>
                        <span
                          class="Transliteration__Joiner"
                        >
                          <span
                            class="Transliteration__Joiner--BROKEN_AWAY"
                          >
                            -
                          </span>
                        </span>
                        <span
                          class="Transliteration__Logogram"
                        >
                          <span
                            class="Transliteration__Logogram--BROKEN_AWAY"
                          >
                            <span
                              class="Transliteration__ValueToken"
                            >
                              <span
                                class="Transliteration__ValueToken--BROKEN_AWAY"
                              >
                                SAḪ
                              </span>
                            </span>
                            <sup
                              class="Transliteration__modifier"
                            />
                            <sup
                              class="Transliteration__flag"
                            />
                          </span>
                        </span>
                      </span>
                    </span>
                    <span
                      class="badge badge-danger"
                    >
                      Empty
                    </span>
                  </span>
                  <span
                    class="Transliteration__wordSeparator Transliteration__wordSeparator--AKKADIAN"
                  >
                     
                  </span>
                  <span
                    class="Transliteration__UnknownNumberOfSigns Transliteration__UnknownNumberOfSigns--AKKADIAN"
                  >
                    …
                  </span>
                  <span
                    class="Transliteration__BrokenAway Transliteration__BrokenAway--AKKADIAN"
                  >
                    <span
                      class="Transliteration__BrokenAway--BROKEN_AWAY"
                    >
                      ]
                    </span>
                  </span>
                </td>
                <td>
                   
                </td>
              </tr>
              <tr
                id="line-2"
              >
                <td
                  class="Transliteration__TextLine"
                >
                  <sup>
                    (
                    2
                    )
                  </sup>
                </td>
                <td
                  colspan="1"
                >
                  <span
                    class="lemmatizer__token-wrapper editable"
                    role="button"
                  >
                    <span
                      class="Transliteration__AkkadianWord Transliteration__AkkadianWord--AKKADIAN"
                    >
                      <span
                        class=""
                      >
                        <span
                          class="Transliteration__Reading"
                        >
                          <span
                            class=""
                          >
                            <span
                              class="Transliteration__ValueToken"
                            >
                              <span
                                class=""
                              >
                                kur
                              </span>
                            </span>
                            <sup
                              class="Transliteration__modifier"
                            />
                            <sup
                              class="Transliteration__flag"
                            />
                          </span>
                        </span>
                        <sup
                          class="Transliteration__flag"
                        />
                      </span>
                    </span>
                    <span
                      class="lemmatizer__lemma-preview"
                    >
                      aklu I
                    </span>
                  </span>
                </td>
                <td>
                   
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
